src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -0,0 +1,315 @@
+
+    return {
+      events: formattedEvents,
+      currentStreamVersion: stream.events.length,
+      streamExists: true,
+    };
+  }
+
+  async aggregateStream<State, EventType extends Event>(
+    streamName: StreamName,
+    options: AggregateStreamOptions<State, EventType, number>,
+  ): Promise<AggregateStreamResult<State, number>> {
+    const stream = await this.readStream<EventType>(streamName, options?.read);
+    const state = stream.events.reduce(options.evolve, options.initialState());
+    return {
+      state,
+      currentStreamVersion: stream.currentStreamVersion,
+      streamExists: stream.streamExists,
+    };
+  }
+
+  async appendToStream<EventType extends Event>(
+    streamName: StreamName,
+    events: EventType[],
+    options?: AppendToStreamOptions<number> & {
+      /**
+       * These will be ran after a the events have been successfully appended to
+       * the stream. `appendToStream` will return after every projection is completed.
+       */
+      projections?: Array<
+        (stream: StreamToProject<EventType>) => void | Promise<void>
+      >;
+      /**
+       * Same as `options.projections` but this will run asynchronously.
+       */
+      asyncProjections?: Array<
+        (stream: StreamToProject<EventType>) => void | Promise<void>
+      >;
+    },
+  ): Promise<AppendToStreamResult<number>> {
+    const eventCreateInputs = events.map(this.stringifyEvent);
+
+    let stream = await this.collection.findOne({
+      streamName: { $eq: streamName },
+    });
+    let createdNewStream = false;
+
+    if (!stream) {
+      const result = await this.collection.insertOne({
+        streamName,
+        events: [],
+        createdAt: new Date(),
+        updatedAt: new Date(),
+      });
+      stream = await this.collection.findOne({
+        _id: result.insertedId,
+      });
+      createdNewStream = true;
+    }
+
+    // TODO: better error here, should rarely happen if ever
+    // if another error was not thrown before this
+    if (!stream) throw new Error('Failed to create stream');
+
+    assertExpectedVersionMatchesCurrent(
+      stream.events.length,
+      options?.expectedStreamVersion,
+      MongoDBEventStoreDefaultStreamVersion,
+    );
+
+    const updatedStream = await this.collection.findOneAndUpdate(
+      {
+        streamName: { $eq: streamName },
+        events: { $size: stream.events.length },
+        updatedAt: new Date(),
+      },
+      { $push: { events: { $each: eventCreateInputs } } },
+      { returnDocument: 'after' },
+    );
+
+    if (!updatedStream) {
+      const currentStream = await this.collection.findOne({
+        streamName: { $eq: streamName },
+      });
+      throw new ExpectedVersionConflictError(
+        currentStream?.events.length ?? -1,
+        stream.events.length,
+      );
+    }
+
+    const { streamType, entityId } = fromStreamName(streamName);
+
+    if (options?.projections) {
+      await Promise.all(
```

Comment:
Executed in no particular order, I don't think there would be a need to have ordered projections.

Reply:
Or at least not a need that couldn't be handled separately as an edge case by the dev

Reply:
I think that's safer to run them sequentially as default and make it configurable to run them in parallel. As you wrote, the general practice is that projections should not depend on each other, but sometimes people does that, so following the least-surprise principle, I'd suggest as in the first sentence.

Reply:
I'll be working on moving the projection registration to be options on the event store instead of per - append call so I'll probably follow the postgres example for this as well.